pennylane/templates/state_preparations.py

```diff
@@ -0,0 +1,315 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+State Preparations
+==================
+
+**Module name:** :mod:`pennylane.templates.state_preparations`
+
+.. currentmodule:: pennylane.templates.state_preparations
+
+This module provides routines that prepare a given state using only
+elementary gates.
+
+Qubit architectures
+-------------------
+
+.. autosummary::
+
+    BasisStatePreparation
+    MottonenStatePreparation
+
+Code details
+^^^^^^^^^^^^
+"""
+import math
+from collections.abc import Iterable
+from scipy import sparse
+
+import numpy as np
+import pennylane as qml
+
+# pylint: disable=len-as-condition,arguments-out-of-order
+
+
+def gray_code(rank):
+    """Generates the Gray code of given rank.
+
+    Args:
+        rank (int): rank of the Gray code (i.e. number of bits)
+    """
+    return ["{0:0{1}b}".format(i ^ (i >> 1), rank) for i in range(0, 1 << rank)]
+
+
+def BasisStatePreparation(basis_state, wires):
+    r"""
+    Prepares a basis state on the given wires using a sequence of Pauli X gates.
+
+    Args:
+        basis_state (array): Input array of shape ``(N,)``, where N is the number of wires
+            the state preparation acts on. ``N`` must be smaller or equal to the total
```

Comment:
Maybe it would nice to simply have ` ``N`` must be equal ` here in the docstring, as the statement on line 69 checks for that.